components/sup/src/cli.rs

```diff
@@ -44,6 +105,224 @@ mod test {
                         "PEER" => ["1.1.1.1", "2.2.2.2"],
                         "PKG_IDENT_OR_ARTIFACT" => "core/redis");
 
+        assert_cli_cmd!(should_handle_permanent_peer,
+                        "hab-sup run --permanent-peer",
+                        "PERMANENT_PEER" => true);
+
+        assert_cli_cmd!(permanent_peer_should_have_short_flag,
+                        "hab-sup run -I",
+                        "PERMANENT_PEER" => true);
+
+        assert_cli_cmd!(should_handle_peer_watch_file,
+                        "hab-sup run --peer-watch-file foobar",
+                        "PEER_WATCH_FILE" => "foobar");
+
+        #[test]
+        fn peer_watch_file_conflicts_with_peer() {
+            let cmd_vec = Vec::from_iter(
+                "hab-sup run --peer-watch-file foobar --peer 1.1.1.1:1111".split_whitespace(),
+            );
+            assert!(cli().get_matches_from_safe(cmd_vec).is_err());
+        }
+
+        assert_cli_cmd!(should_handle_ring,
+                        "hab-sup run --ring fizbang",
+                        "RING" => "fizbang");
+
+        assert_cli_cmd!(ring_should_have_short_flag,
+                        "hab-sup run -r flippitygibbit",
+                        "RING" => "flippitygibbit");
+
+        #[test]
+        fn ring_conflicts_with_ring_key() {
+            let cmd_vec =
+                Vec::from_iter("hab-sup run --ring myring --ring-key foobar".split_whitespace());
+            assert!(cli().get_matches_from_safe(cmd_vec).is_err());
+        }
+
+        assert_cli_cmd!(should_handle_ring_key,
+                        "hab-sup run --ring-key secrit",
+                        "RING_KEY" => "secrit");
+
+        assert_cli_cmd!(should_handle_channel,
+                        "hab-sup run --channel dev",
+                        "CHANNEL" => "dev");
+
+        assert_cli_cmd!(should_handle_bldr_url,
+                        "hab-sup run --url http://not.a.real.com",
+                        "BLDR_URL" => "http://not.a.real.com");
+
+        assert_cli_cmd!(bldr_url_should_have_a_short_flag,
+                        "hab-sup run -u http://not.a.real.com",
+                        "BLDR_URL" => "http://not.a.real.com");
+
+        assert_cli_cmd!(bldr_url_should_have_a_default_value,
+                        "hab-sup run",
+                        "BLDR_URL" => "https://bldr.habitat.sh");
+
+        #[test]
+        fn invalid_bldr_url_should_fail() {
+            let cmd_vec = Vec::from_iter("hab-sup run -u bad-url".split_whitespace());
```

Comment:
Similar to the addr args, a more plausibly bad URL would be a more valuable test:
```suggestion
            let cmd_vec = Vec::from_iter("hab-sup run -u https:/bldr.habitat.sh".split_whitespace());
```
I believe this example actually *does* pass our validation indicating we may want to do something a bit stricter than `Url::parse`, depending on how we're consuming that value.

Reply:
I agree that it seems like `Url::parse` is a bit too lenient, but I don't want to tackle that in this PR. Would prefer to keep this one relatively focused on capturing the existing behavior.

Reply:
That's fair; please file an issue for the follow-up.